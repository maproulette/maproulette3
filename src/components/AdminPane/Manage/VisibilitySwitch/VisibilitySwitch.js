import React, { Component } from 'react'
import PropTypes from 'prop-types'
import WithChallengeManagement
       from '../../HOCs/WithChallengeManagement/WithChallengeManagement'


/**
 * VisibilitySwitch renders a simple switch, with optional label, that
 * toggles the given challenge's enabled status true or false as the
 * switch is activated or deactivated, respectively.
 *
 * @author [Neil Rotstan](https://github.com/nrotstan)
 */
export class VisibilitySwitch extends Component {
  toggleVisible = () => {
    this.props.updateEnabled(this.props.challenge.id,
                             !this.props.challenge.enabled)
  }

  render() {
    if (!this.props.challenge) {
      return null
    }

    return (
      <div className="field visibility-switch" onClick={this.toggleVisible}>
        <input type="checkbox" className="switch is-rounded short-and-wide"
               disabled={this.props.disabled}
               checked={this.props.challenge.enabled}
               onChange={() => null} />
        <label>{this.props.label}</label>
      </div>
    )
  }
}

VisibilitySwitch.propTypes = {
  challenge: PropTypes.object,
  updateEnabled: PropTypes.func.isRequired,
  label: PropTypes.node,
}

export default WithChallengeManagement(VisibilitySwitch)
