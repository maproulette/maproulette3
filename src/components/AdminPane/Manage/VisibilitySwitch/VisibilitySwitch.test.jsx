import { describe, expect, it, vi } from "vitest";
import "@testing-library/jest-dom";
import { fireEvent } from "@testing-library/react";
import { VisibilitySwitch } from "./VisibilitySwitch";

describe("Visibility Switch", () => {
  it("renders Visibility Switch if a challenge object is provided", () => {
    const { container } = global.withProvider(
      <VisibilitySwitch updateEnabled={() => updateEnabled()} challenge={{ enabled: false }} />,
    );

    expect(container.firstChild.classList["0"]).toBe("mr-flex");
  });

  it("input onClick event does not call updateEnabled", () => {
    const updateEnabled = vi.fn();
    const { container } = global.withProvider(
      <VisibilitySwitch updateEnabled={() => updateEnabled()} challenge={{ enabled: false }} />,
    );

    const element = container.querySelector("input[type=checkbox]");

    fireEvent.click(element);

    expect(updateEnabled).toHaveBeenCalledTimes(0);
  });
});
