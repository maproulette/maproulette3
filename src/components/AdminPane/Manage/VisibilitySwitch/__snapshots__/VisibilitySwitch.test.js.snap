// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows visibility switch 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <VisibilitySwitch
    challenge={
      Object {
        "enabled": false,
        "id": 123,
      }
    }
    updateEnabled={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          checked={false}
          className="switch is-rounded short-and-wide"
          onChange={[Function]}
          type="checkbox"
        />,
        <label />,
      ],
      "className": "field visibility-switch",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "checked": false,
          "className": "switch is-rounded short-and-wide",
          "disabled": undefined,
          "onChange": [Function],
          "type": "checkbox",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": "label",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            checked={false}
            className="switch is-rounded short-and-wide"
            onChange={[Function]}
            type="checkbox"
          />,
          <label />,
        ],
        "className": "field visibility-switch",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "checked": false,
            "className": "switch is-rounded short-and-wide",
            "disabled": undefined,
            "onChange": [Function],
            "type": "checkbox",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "label",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`switch state is off when challenge not enabled 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <VisibilitySwitch
    challenge={
      Object {
        "enabled": false,
        "id": 123,
      }
    }
    updateEnabled={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          checked={false}
          className="switch is-rounded short-and-wide"
          onChange={[Function]}
          type="checkbox"
        />,
        <label />,
      ],
      "className": "field visibility-switch",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "checked": false,
          "className": "switch is-rounded short-and-wide",
          "disabled": undefined,
          "onChange": [Function],
          "type": "checkbox",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": "label",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            checked={false}
            className="switch is-rounded short-and-wide"
            onChange={[Function]}
            type="checkbox"
          />,
          <label />,
        ],
        "className": "field visibility-switch",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "checked": false,
            "className": "switch is-rounded short-and-wide",
            "disabled": undefined,
            "onChange": [Function],
            "type": "checkbox",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "label",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`switch state is on when challenge is enabled 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <VisibilitySwitch
    challenge={
      Object {
        "enabled": true,
        "id": 123,
      }
    }
    updateEnabled={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          checked={true}
          className="switch is-rounded short-and-wide"
          onChange={[Function]}
          type="checkbox"
        />,
        <label />,
      ],
      "className": "field visibility-switch",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "checked": true,
          "className": "switch is-rounded short-and-wide",
          "disabled": undefined,
          "onChange": [Function],
          "type": "checkbox",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": "label",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            checked={true}
            className="switch is-rounded short-and-wide"
            onChange={[Function]}
            type="checkbox"
          />,
          <label />,
        ],
        "className": "field visibility-switch",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "checked": true,
            "className": "switch is-rounded short-and-wide",
            "disabled": undefined,
            "onChange": [Function],
            "type": "checkbox",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "label",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
